go/src/meguca/geoip/db.go

```diff
@@ -4,38 +4,93 @@ import (
 	"net"
 	"strings"
 	"time"
+	"os"
+	"errors"
+	"fmt"
+	"sync"
+	"io/ioutil"
+	"net/http"
+	"crypto/md5"
+	"path/filepath"
+
+	"meguca/db"
 
 	"github.com/oschwald/maxminddb-golang"
 	"github.com/go-playground/log"
+	"github.com/mholt/archiver"
 )
 
 // nil, if database not loaded
-var db *maxminddb.Reader
+var (
+	// Ensure we only load the old GeoIP DB once on server start
+	once sync.Once
+	// Ensures no data races
+	rw sync.RWMutex
+	// GeoIP database
+	gdb *maxminddb.Reader
+)
 
+// NY location
 var NY *time.Location
 
 func init() {
 	NY, _ = time.LoadLocation("America/New_York")
 }
 
+// Load checks if the GeoLite DB exists, and calls load it if it does
 func Load() error {
-	var err error
-	db, err = maxminddb.Open("GeoLite2-Country.mmdb")
+	go func() {
+		if err := check(); err != nil {
+			rw.Lock()
+			defer rw.Unlock()
+			gdb = nil
+			log.Warn("Unable to use GeoLite DB: ", err)
+		}
+	}()
+
+	return nil
+}
+
+// loads the GeoLite DB
+func load() (err error) {
+	rw.Lock()
+	defer rw.Unlock()
+	gdb, err = maxminddb.Open("GeoLite2-Country.mmdb")
+	return
+}
+
+// reloads the GeoLite DB
+func reload() error {
+	rw.Lock()
+	defer rw.Unlock()
+
+	if gdb == nil {
+		return load()
+	}
+
+	err := gdb.Close()
+
 	if err != nil {
-		db = nil
+		return err
 	}
-	return nil
+
+	gdb = nil
+	return load()
```

Comment:
If this is called, a double lock (deadlock) will occur, because rw will only only unlock, when reload() returns and load() will try to acquire a lock. You can avoid this by moving 
```go
	if gdb == nil {
		return load()
	}
 	err := gdb.Close()
```
to load() and thereby removing the need for reload() even existing.  

Yeah, I need to be more careful in reviewing your locking code.